### PR TITLE
Enhancement: Run jobs on GitHub Actions

### PR DESCRIPTION
**Pull request by localheinz**
This PR

* [x] switches the build process from Travis CI to GitHub Actions

💁‍♂ For reference, see

- https://github.com/localheinz/composer/pull/1
- https://github.com/localheinz/composer/pull/1/checks

### CODE REVIEW

**Comment by localheinz** [comment]
As far as I can tell, this works fine.

File: .github/workflows/continuous-integration.yml [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+      - name: "Prepare git environment"
+        run: "git config --global user.name composer && git config --global user.email composer@example.com"
+
+      - name: "Set SYMFONY_PHPUNIT_VERSION environment variable"
+        if: "matrix.php-version == '7.4'"
+        run: "echo \"::set-env name=SYMFONY_PHPUNIT_VERSION::7.5\""
+
+      - name: "Run tests"
+        run: "vendor/bin/simple-phpunit"
```
Comment (staabm) [inline .github/workflows/continuous-integration.yml]: what do you think about activating a phpunit problem matcher?
https://github.com/shivammathur/setup-php#problem-matchers

Comment (stof) [reply .github/workflows/continuous-integration.yml]: Given that problem matchers are limited to adding 10 annotations per step, with a limit of 50 per job, I find them a bit overrated actually.

Comment (staabm) [reply .github/workflows/continuous-integration.yml]: still its better then looking up errors in the logfiles (and saves time)

File: .github/workflows/continuous-integration.yml [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+      - name: "Cache dependencies installed with composer"
+        uses: "actions/cache@v1"
+        with:
+          path: "${{ steps.determine-composer-cache-directory.outputs.directory }}"
+          key: "php-${{ matrix.php-version }}-symfony-php-unit-version-${{ env.SYMFONY_PHPUNIT_VERSION }}-${{ hashFiles('**/composer.lock') }}"
+          restore-keys: "php-${{ matrix.php-version }}-symfony-php-unit-version-${{ env.SYMFONY_PHPUNIT_VERSION }}"
+
+      - name: "Install highest dependencies from composer.json"
+        if: "matrix.php-version == '7.4'"
```
Comment (stof) [inline .github/workflows/continuous-integration.yml]: shouldn't we still keep a job running on PHP 7.4 with our locked dependencies ?

File: .travis.yml [hunk 3]
```diff
@@ -16,20 +16,7 @@ addons:
...
-    - php: 5.3
-      dist: precise
-    - php: 5.4
-    - php: 5.5
-    - php: 5.6
-    - php: 7.0
-    - php: 7.1
-    - php: 7.2
-    - php: 7.3
     - php: nightly
```
Comment (stof) [inline .travis.yml]: keeping a single job, which is in `allow_failures` looks weird to me.

File: .travis.yml [hunk 4]
```diff
@@ -16,20 +16,7 @@ addons:
...
 matrix:
   include:
-    - php: 5.3
-      dist: precise
-    - php: 5.4
-    - php: 5.5
-    - php: 5.6
-    - php: 7.0
-    - php: 7.1
-    - php: 7.2
```
Comment (stof) [inline .travis.yml]: This PHP 7.2 job is the one uploading the composer.phar to the github release on tags. So this is now lost as the `deploy` section will never match anymore on Travis and you don't have the equivalent on github actions.

File: .github/workflows/continuous-integration.yml (hunk 1 above)
Comment (stof) [reply .github/workflows/continuous-integration.yml]: but as soon as you have more than 10 errors, you must go in the log files (and there is nothing telling you that there might have additional errors in the log file)

File: .travis.yml (hunk 4 above)
Comment (shivammathur) [reply .travis.yml]: @localheinz @stof we can have another workflow to deploy, similar to [this](https://github.com/staabm/annotate-pull-request-from-checkstyle/blob/master/.github/workflows/continuous-deployment.yml) for composer

File: .github/workflows/continuous-integration.yml (hunk 2 above)
Comment (localheinz) [reply .github/workflows/continuous-integration.yml]: We currently aren’t running such a job on Travis.

File: .travis.yml (hunk 3 above)
Comment (localheinz) [reply .travis.yml]: We can't run a job in GitHub Actions and allow it to fail at the some time.

We could use [`continue-on-error`](https://help.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idstepscontinue-on-error), but I don't think that running jobs that are allowed to fail makes sense at all.

File: .travis.yml (hunk 4 above)
Comment (localheinz) [reply .travis.yml]: Releases can be created with [`actions/create-release`](https://github.com/actions/create-release). Release assets can be uploaded using [`actions/upload-release-asset`](https://github.com/actions/upload-release-asset).

File: .github/workflows/continuous-integration.yml (hunk 2 above)
Comment (Seldaek) [reply .github/workflows/continuous-integration.yml]: That would probably make sense to have though if you don't mind adding it here.

**Comment by Seldaek** [comment]
This seems to trigger two builds for every PR (one for push and one for pull_request), isn't this a little wasteful? Not sure what can be done about it..

**Comment by stof** [comment]
@Seldaek when the push in the main repo, that's also what Travis does (building the branch, and building the PR which runs on the merge commit of that branch with master).
Both Travis and Github Actions allow to restrict push builds to pushes to some specific branches. But that is not used on Travis currently either.

**Comment by Seldaek** [comment]
Ok no problem about that then. Adding a regular 7.4 build would be nice still.

**Comment by Seldaek** [comment]
Merged in 1.10/master, last commit shows nothing from github tho, and https://github.com/composer/composer/actions also empty.. Sorry I'm probably dumb here but I can't find out how to enable this, only how to create a new file?!

Also #8773 should be backported here if possible, as a single build in the matrix.. not sure if we can avoid doing locked/latest builds for every php version too, that seems a little extreme.

**Comment by staabm** [comment]
@Seldaek actions are activated in the repo config?

grafik

**Comment by Seldaek** [comment]
Yeah I checked that and it's on.

**Comment by Seldaek** [comment]
Ah it seems it builds on https://github.com/composer/composer/pull/8803 so maybe it's only enabled for PRs now?

**Comment by staabm** [comment]
> Ah it seems it builds on #8803

yep

https://github.com/composer/composer/blob/1eaa8cd55fbcc6e5fbee10820456da3700f46593/.github/workflows/continuous-integration.yml#L6

fix incoming.

**Comment by Seldaek** [comment]
Migrated all the travis/appveyor  stuff we still had on github actions now. Thanks for getting the ball rolling @localheinz

**Comment by localheinz** [comment]
@Seldaek 

You are welcome!

Feel free to ping me any time when I can help!

**Pull request closed without merging**